### PR TITLE
Move check for layer construction failure before first use

### PR DESCRIPTION
**Pull request by ndryden**
Otherwise if construction fails, we segfault.

### CODE REVIEW

File: src/proto/factories/layer_graph_factory.cpp [hunk 1]
```diff
@@ -172,6 +172,11 @@ construct_layer_graph(lbann_comm* comm,
 
 #undef TEMPLATE_INSTANTIATION
 
+    // Check that layer has been constructed
```
Comment (tbennun) [inline src/proto/factories/layer_graph_factory.cpp:175]: Not sure about this part

File: src/proto/factories/layer_graph_factory.cpp [hunk 2]
```diff
@@ -172,6 +172,11 @@ construct_layer_graph(lbann_comm* comm,
...
+    if (l == nullptr) {
+      LBANN_ERROR("Could not construct layer ", name);
```
Comment (tbennun) [inline src/proto/factories/layer_graph_factory.cpp:177]: Why is that?

File: src/proto/factories/layer_graph_factory.cpp [hunk 3]
```diff
@@ -188,10 +193,6 @@ construct_layer_graph(lbann_comm* comm,
     ps.depth_groups = proto_layer.parallel_strategy().depth_groups();
     ps.depth_splits = proto_layer.parallel_strategy().depth_splits();
 
-    // Check that layer has been constructed
-    if (l == nullptr)
-      LBANN_ERROR("Could not construct layer ", name);
-
```
Comment (tbennun) [inline src/proto/factories/layer_graph_factory.cpp:194]: Might want to check that a second time here too, just in case.

**Review by tbennun — CHANGES_REQUESTED** [review]

File: src/proto/factories/layer_graph_factory.cpp (hunk 3 above)
Comment (ndryden) [reply src/proto/factories/layer_graph_factory.cpp:194]: Why? The pointer can't change at this point.

File: src/proto/factories/layer_graph_factory.cpp (hunk 2 above)
Comment (ndryden) [reply src/proto/factories/layer_graph_factory.cpp:177]: If it's null, it means the instantiation attempts (generated by macros) failed, as the default value of `l` is `nullptr`.

**Review by tbennun — APPROVED** [review]

**Comment by ndryden** [comment]
Thanks @tbennun, I know this was a complicated PR.

**Pull request merged**